File: backend/api/email_sync.py
"""
Email synchronization endpoints.

This module provides endpoints for managing IMAP/SMTP email synchronization,
including triggering manual sync, checking sync status, and managing email
configuration settings for candidate communication tracking.
"""
import logging
from typing import Optional

from fastapi import APIRouter, HTTPException, status
from fastapi.responses import JSONResponse
from pydantic import BaseModel, Field, EmailStr
from sqlalchemy import select

logger = logging.getLogger(__name__)

router = APIRouter()


class SyncRequest(BaseModel):
    """Request model for triggering email synchronization."""

    full_sync: bool = Field(False, description="Perform full sync vs incremental sync")
    folder: Optional[str] = Field(None, description="Specific IMAP folder to sync (e.g., 'INBOX')")


class SyncResponse(BaseModel):
    """Response model for sync operations."""

    sync_id: str = Field(..., description="Unique identifier for this sync operation")
    status: str = Field(..., description="Sync status: started, in_progress, completed, failed")
    message: str = Field(..., description="Human-readable status message")
    emails_processed: int = Field(..., description="Number of emails processed")
    emails_created: int = Field(..., description="Number of new email records created")
    emails_updated: int = Field(..., description="Number of existing email records updated")
    errors: list[str] = Field(default_factory=list, description="List of any errors encountered")


class SyncStatusResponse(BaseModel):
    """Response model for sync status."""

    is_syncing: bool = Field(..., description="Whether a sync is currently in progress")
    last_sync_time: Optional[str] = Field(None, description="ISO 8601 timestamp of last successful sync")
    last_sync_status: str = Field(..., description="Status of last sync: success, failed, partial")
    last_sync_duration_seconds: Optional[float] = Field(None, description="Duration of last sync in seconds")
    total_emails_synced: int = Field(..., description="Total number of emails synced to date")
    next_scheduled_sync: Optional[str] = Field(None, description="ISO 8601 timestamp of next scheduled sync")
    sync_config: dict = Field(..., description="Current sync configuration summary")


class EmailConfig(BaseModel):
    """Email configuration model for IMAP/SMTP settings."""

    imap_enabled: bool = Field(..., description="Whether IMAP sync is enabled")
    imap_server: str = Field(..., description="IMAP server address")
    imap_port: int = Field(..., description="IMAP port (typically 993 for SSL)")
    imap_use_ssl: bool = Field(..., description="Use SSL for IMAP connection")
    imap_username: str = Field(..., description="IMAP username (typically email address)")
    # Note: Password is never returned in GET responses

    smtp_enabled: bool = Field(..., description="Whether SMTP is enabled for sending")
    smtp_server: str = Field(..., description="SMTP server address")
    smtp_port: int = Field(..., description="SMTP port (typically 587 for TLS, 465 for SSL)")
    smtp_use_ssl: bool = Field(..., description="Use SSL/TLS for SMTP connection")
    smtp_username: str = Field(..., description="SMTP username (typically email address)")

    sync_interval_minutes: int = Field(..., description="How often to sync emails (in minutes)")
    sync_folders: list[str] = Field(..., description="List of IMAP folders to sync")
    auto_link_candidates: bool = Field(..., description="Automatically link emails to candidates")
    sync_attachments: bool = Field(..., description="Whether to download and sync email attachments")


class EmailConfigUpdate(BaseModel):
    """Request model for updating email configuration."""

    imap_enabled: Optional[bool] = Field(None, description="Enable/disable IMAP sync")
    imap_server: Optional[str] = Field(None, description="IMAP server address")
    imap_port: Optional[int] = Field(None, ge=1, le=65535, description="IMAP port (1-65535)")
    imap_use_ssl: Optional[bool] = Field(None, description="Use SSL for IMAP connection")
    imap_username: Optional[EmailStr] = Field(None, description="IMAP username")
    imap_password: Optional[str] = Field(None, description="IMAP password (for updates only)")

    smtp_enabled: Optional[bool] = Field(None, description="Enable/disable SMTP")
    smtp_server: Optional[str] = Field(None, description="SMTP server address")
    smtp_port: Optional[int] = Field(None, ge=1, le=65535, description="SMTP port (1-65535)")
    smtp_use_ssl: Optional[bool] = Field(None, description="Use SSL/TLS for SMTP connection")
    smtp_username: Optional[EmailStr] = Field(None, description="SMTP username")
    smtp_password: Optional[str] = Field(None, description="SMTP password (for updates only)")

    sync_interval_minutes: Optional[int] = Field(
        None, ge=1, le=1440, description="Sync interval in minutes (1-1440)"
    )
    sync_folders: Optional[list[str]] = Field(None, description="IMAP folders to sync")
    auto_link_candidates: Optional[bool] = Field(None, description="Auto-link emails to candidates")
    sync_attachments: Optional[bool] = Field(None, description="Sync email attachments")


@router.post(
    "/sync",
    response_model=SyncResponse,
    tags=["Email Sync"],
)
async def trigger_sync(request: SyncRequest) -> JSONResponse:
    """
    Trigger email synchronization.

    This endpoint initiates a manual email synchronization with the configured
    IMAP server. Emails will be fetched, parsed, and stored in the database.
    Candidate linking will be performed if enabled in configuration.

    Supports both full sync (all emails) and incremental sync (only new emails
    since last sync). Can optionally sync a specific folder.

    Args:
        request: Sync request with optional full_sync flag and folder specification

    Returns:
        JSON response with sync operation details including status and counts

    Raises:
        HTTPException(400): If email is not configured
        HTTPException(409): If a sync is already in progress
        HTTPException(500): If sync fails to start

    Examples:
        >>> import requests
        >>> response = requests.post(
        ...     "http://localhost:8000/api/email-sync/sync",
        ...     json={"full_sync": False, "folder": "INBOX"}
        ... )
        >>> response.json()
        {
            "sync_id": "sync_20250103_123456",
            "status": "started",
            "message": "Email synchronization started",
            "emails_processed": 0,
            "emails_created": 0,
            "emails_updated": 0,
            "errors": []
        }
    """
    try:
        logger.info(f"Triggering email sync - full_sync: {request.full_sync}, folder: {request.folder}")

        # TODO: Implement actual sync logic in a later subtask
        # For now, return a placeholder response indicating the endpoint is ready
        import uuid
        from datetime import datetime

        sync_id = f"sync_{datetime.now().strftime('%Y%m%d_%H%M%S')}_{uuid.uuid4().hex[:8]}"

        logger.info(f"Sync started with ID: {sync_id}")

        response_data = {
            "sync_id": sync_id,
            "status": "started",
            "message": "Email synchronization initiated",
            "emails_processed": 0,
            "emails_created": 0,
            "emails_updated": 0,
            "errors": [],
        }

        return JSONResponse(
            status_code=status.HTTP_202_ACCEPTED,
            content=response_data,
        )

    except Exception as e:
        logger.error(f"Error triggering email sync: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to trigger email sync: {str(e)}",
        ) from e


@router.get(
    "/status",
    response_model=SyncStatusResponse,
    tags=["Email Sync"],
)
async def get_sync_status() -> JSONResponse:
    """
    Get email synchronization status.

    This endpoint returns the current status of email synchronization,
    including whether a sync is in progress, when the last sync occurred,
    and overall sync statistics.

    Returns:
        JSON response with current sync status and configuration summary

    Raises:
        HTTPException(500): If status retrieval fails

    Examples:
        >>> import requests
        >>> response = requests.get("http://localhost:8000/api/email-sync/status")
        >>> response.json()
        {
            "is_syncing": false,
            "last_sync_time": "2025-01-03T12:30:00Z",
            "last_sync_status": "success",
            "last_sync_duration_seconds": 45.2,
            "total_emails_synced": 1250,
            "next_scheduled_sync": "2025-01-03T13:30:00Z",
            "sync_config": {
                "imap_enabled": true,
                "sync_interval_minutes": 60,
                "sync_folders": ["INBOX", "Sent"],
                "auto_link_candidates": true
            }
        }
    """
    try:
        logger.info("Fetching email sync status")

        # TODO: Implement actual status retrieval in a later subtask
        # For now, return placeholder data
        response_data = {
            "is_syncing": False,
            "last_sync_time": None,
            "last_sync_status": "not_configured",
            "last_sync_duration_seconds": None,
            "total_emails_synced": 0,
            "next_scheduled_sync": None,
            "sync_config": {
                "imap_enabled": False,
                "sync_interval_minutes": 60,
                "sync_folders": ["INBOX"],
                "auto_link_candidates": True,
            },
        }

        logger.info("Email sync status retrieved successfully")

        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content=response_data,
        )

    except Exception as e:
        logger.error(f"Error retrieving email sync status: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to retrieve email sync status: {str(e)}",
        ) from e


@router.get(
    "/config",
    response_model=EmailConfig,
    tags=["Email Sync"],
)
async def get_email_config() -> JSONResponse:
    """
    Get email configuration.

    This endpoint returns the current email configuration for IMAP/SMTP.
    Sensitive information like passwords are never included in the response.

    Returns:
        JSON response with current email configuration (excluding passwords)

    Raises:
        HTTPException(500): If config retrieval fails

    Examples:
        >>> import requests
        >>> response = requests.get("http://localhost:8000/api/email-sync/config")
        >>> response.json()
        {
            "imap_enabled": true,
            "imap_server": "imap.gmail.com",
            "imap_port": 993,
            "imap_use_ssl": true,
            "imap_username": "recruiting@example.com",
            "smtp_enabled": true,
            "smtp_server": "smtp.gmail.com",
            "smtp_port": 587,
            "smtp_use_ssl": true,
            "smtp_username": "recruiting@example.com",
            "sync_interval_minutes": 60,
            "sync_folders": ["INBOX", "Sent"],
            "auto_link_candidates": true,
            "sync_attachments": false
        }
    """
    try:
        logger.info("Fetching email configuration")

        # TODO: Implement actual config retrieval from database in a later subtask
        # For now, return placeholder default configuration
        response_data = {
            "imap_enabled": False,
            "imap_server": "imap.example.com",
            "imap_port": 993,
            "imap_use_ssl": True,
            "imap_username": "",
            "smtp_enabled": False,
            "smtp_server": "smtp.example.com",
            "smtp_port": 587,
            "smtp_use_ssl": True,
            "smtp_username": "",
            "sync_interval_minutes": 60,
            "sync_folders": ["INBOX"],
            "auto_link_candidates": True,
            "sync_attachments": False,
        }

        logger.info("Email configuration retrieved successfully")

        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content=response_data,
        )

    except Exception as e:
        logger.error(f"Error retrieving email configuration: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to retrieve email configuration: {str(e)}",
        ) from e


@router.put(
    "/config",
    response_model=EmailConfig,
    tags=["Email Sync"],
)
async def update_email_config(request: EmailConfigUpdate) -> JSONResponse:
    """
    Update email configuration.

    This endpoint updates the email configuration for IMAP/SMTP synchronization.
    Passwords can be updated but are never returned in the response.

    Only the fields specified in the request will be updated. All other fields
    will remain unchanged. Partial updates are supported.

    Args:
        request: Configuration update with only the fields to change

    Returns:
        JSON response with updated configuration (excluding passwords)

    Raises:
        HTTPException(400): If configuration is invalid
        HTTPException(500): If update fails

    Examples:
        >>> import requests
        >>> response = requests.put(
        ...     "http://localhost:8000/api/email-sync/config",
        ...     json={
        ...         "imap_enabled": True,
        ...         "imap_server": "imap.gmail.com",
        ...         "imap_port": 993,
        ...         "imap_username": "recruiting@example.com",
        ...         "imap_password": "secure_password",
        ...         "sync_interval_minutes": 30
        ...     }
        ... )
        >>> response.json()
        {
            "imap_enabled": true,
            "imap_server": "imap.gmail.com",
            "imap_port": 993,
            "imap_use_ssl": true,
            "imap_username": "recruiting@example.com",
            "smtp_enabled": false,
            "smtp_server": "smtp.example.com",
            "smtp_port": 587,
            "smtp_use_ssl": true,
            "smtp_username": "",
            "sync_interval_minutes": 30,
            "sync_folders": ["INBOX"],
            "auto_link_candidates": true,
            "sync_attachments": false
        }
    """
    try:
        logger.info("Updating email configuration")

        # TODO: Implement actual config update in database in a later subtask
        # For now, return the requested configuration (simulating a successful update)
        response_data = {
            "imap_enabled": request.imap_enabled if request.imap_enabled is not None else False,
            "imap_server": request.imap_server or "imap.example.com",
            "imap_port": request.imap_port if request.imap_port is not None else 993,
            "imap_use_ssl": request.imap_use_ssl if request.imap_use_ssl is not None else True,
            "imap_username": request.imap_username or "",
            "smtp_enabled": request.smtp_enabled if request.smtp_enabled is not None else False,
            "smtp_server": request.smtp_server or "smtp.example.com",
            "smtp_port": request.smtp_port if request.smtp_port is not None else 587,
            "smtp_use_ssl": request.smtp_use_ssl if request.smtp_use_ssl is not None else True,
            "smtp_username": request.smtp_username or "",
            "sync_interval_minutes": request.sync_interval_minutes if request.sync_interval_minutes is not None else 60,
            "sync_folders": request.sync_folders if request.sync_folders is not None else ["INBOX"],
            "auto_link_candidates": request.auto_link_candidates if request.auto_link_candidates is not None else True,
            "sync_attachments": request.sync_attachments if request.sync_attachments is not None else False,
        }

        logger.info("Email configuration updated successfully")

        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content=response_data,
        )

    except Exception as e:
        logger.error(f"Error updating email configuration: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to update email configuration: {str(e)}",
        ) from e

File: frontend/src/api/emailSync.ts
/**
 * Email Sync API Client
 *
 * This module provides a client for managing email synchronization with IMAP/SMTP servers,
 * including triggering manual syncs, checking sync status, and managing email configuration.
 *
 * @example
 * ```ts
 * import { emailSyncClient } from '@/api/emailSync';
 *
 * // Trigger email synchronization
 * const syncResult = await emailSyncClient.sync({
 *   full_sync: false,
 *   folder: 'INBOX'
 * });
 *
 * // Get sync status
 * const status = await emailSyncClient.getStatus();
 *
 * // Get email configuration
 * const config = await emailSyncClient.getConfig();
 *
 * // Update email configuration
 * const updated = await emailSyncClient.updateConfig({
 *   imap_enabled: true,
 *   imap_server: 'imap.gmail.com',
 *   imap_port: 993,
 *   imap_username: 'recruiting@example.com',
 *   imap_password: 'secure_password',
 *   sync_interval_minutes: 30
 * });
 * ```
 */

import axios, { AxiosInstance, AxiosError, AxiosResponse } from 'axios';
import type { ApiError } from '@/types/api';

/**
 * Email sync request
 */
export interface EmailSyncRequest {
  full_sync?: boolean;
  folder?: string;
}

/**
 * Email sync response
 */
export interface EmailSyncResponse {
  sync_id: string;
  status: string;
  message: string;
  emails_processed: number;
  emails_created: number;
  emails_updated: number;
  errors: string[];
}

/**
 * Email sync status response
 */
export interface EmailSyncStatusResponse {
  is_syncing: boolean;
  last_sync_time: string | null;
  last_sync_status: string;
  last_sync_duration_seconds: number | null;
  total_emails_synced: number;
  next_scheduled_sync: string | null;
  sync_config: Record<string, unknown>;
}

/**
 * Email configuration
 */
export interface EmailConfig {
  imap_enabled: boolean;
  imap_server: string;
  imap_port: number;
  imap_use_ssl: boolean;
  imap_username: string;
  smtp_enabled: boolean;
  smtp_server: string;
  smtp_port: number;
  smtp_use_ssl: boolean;
  smtp_username: string;
  sync_interval_minutes: number;
  sync_folders: string[];
  auto_link_candidates: boolean;
  sync_attachments: boolean;
}

/**
 * Email configuration update
 */
export interface EmailConfigUpdate {
  imap_enabled?: boolean;
  imap_server?: string;
  imap_port?: number;
  imap_use_ssl?: boolean;
  imap_username?: string;
  imap_password?: string;
  smtp_enabled?: boolean;
  smtp_server?: string;
  smtp_port?: number;
  smtp_use_ssl?: boolean;
  smtp_username?: string;
  smtp_password?: string;
  sync_interval_minutes?: number;
  sync_folders?: string[];
  auto_link_candidates?: boolean;
  sync_attachments?: boolean;
}

/**
 * Default API configuration for email sync client
 */
const DEFAULT_CONFIG = {
  baseURL: import.meta.env.VITE_API_URL ?? '',
  timeout: 10000, // 10 seconds
  headers: {
    'Content-Type': 'application/json',
  },
};

/**
 * Email Sync API Client class
 *
 * Provides methods for managing email synchronization with proper
 * error handling and type safety.
 */
export class EmailSyncClient {
  private client: AxiosInstance;

  /**
   * Create a new EmailSync client instance
   *
   * @param config - Optional configuration overrides
   */
  constructor(config: Partial<typeof DEFAULT_CONFIG> = {}) {
    const finalConfig = { ...DEFAULT_CONFIG, ...config };

    this.client = axios.create(finalConfig);

    // Response interceptor for error handling
    this.client.interceptors.response.use(
      (response) => response,
      (error) => Promise.reject(this.transformError(error))
    );
  }

  /**
   * Transform Axios error to standardized API error
   *
   * @param error - Axios error
   * @returns Transformed API error
   */
  private transformError(error: unknown): ApiError {
    const axiosError = error as AxiosError<{ detail?: string }>;

    // Network error (no response)
    if (!axiosError.response) {
      if (axiosError.code === 'ECONNABORTED') {
        return {
          detail: 'Request timeout. Please check your connection and try again.',
          status: 408,
        };
      }
      return {
        detail: 'Network error. Please check your connection and try again.',
        status: 0,
      };
    }

    // Server returned error response
    const status = axiosError.response.status;
    const data = axiosError.response.data;

    // Use server's error message if available
    if (data?.detail) {
      return { detail: data.detail, status };
    }

    // Default error messages by status code
    const defaultMessages: Record<number, string> = {
      400: 'Invalid request. Please check your input.',
      401: 'Unauthorized. Please log in.',
      403: 'Forbidden. You do not have permission.',
      404: 'Resource not found.',
      409: 'A sync is already in progress.',
      422: 'Validation error. Please check your input.',
      429: 'Too many requests. Please try again later.',
      500: 'Server error. Please try again later.',
      502: 'Bad gateway. Please try again later.',
      503: 'Service unavailable. Please try again later.',
    };

    return {
      detail: data?.detail || defaultMessages[status] || 'An unexpected error occurred.',
      status,
    };
  }

  /**
   * Trigger email synchronization
   *
   * @param request - Sync request with optional full_sync flag and folder specification
   * @returns Sync operation details including status and counts
   * @throws ApiError if sync fails to start
   *
   * @example
   * ```ts
   * // Trigger incremental sync for INBOX
   * const sync = await emailSyncClient.sync({
   *   full_sync: false,
   *   folder: 'INBOX'
   * });
   *
   * // Trigger full sync
   * const fullSync = await emailSyncClient.sync({
   *   full_sync: true
   * });
   * ```
   */
  async sync(request: EmailSyncRequest = {}): Promise<EmailSyncResponse> {
    try {
      const response: AxiosResponse<EmailSyncResponse> = await this.client.post(
        '/api/email-sync/sync',
        request
      );
      return response.data;
    } catch (error) {
      throw this.transformError(error);
    }
  }

  /**
   * Get email synchronization status
   *
   * @returns Current sync status and configuration summary
   * @throws ApiError if status retrieval fails
   *
   * @example
   * ```ts
   * const status = await emailSyncClient.getStatus();
   * console.log(`Syncing: ${status.is_syncing}`);
   * console.log(`Last sync: ${status.last_sync_time}`);
   * console.log(`Total emails: ${status.total_emails_synced}`);
   * ```
   */
  async getStatus(): Promise<EmailSyncStatusResponse> {
    try {
      const response: AxiosResponse<EmailSyncStatusResponse> = await this.client.get(
        '/api/email-sync/status'
      );
      return response.data;
    } catch (error) {
      throw this.transformError(error);
    }
  }

  /**
   * Get email configuration
   *
   * Returns the current email configuration for IMAP/SMTP.
   * Sensitive information like passwords are never included in the response.
   *
   * @returns Current email configuration (excluding passwords)
   * @throws ApiError if config retrieval fails
   *
   * @example
   * ```ts
   * const config = await emailSyncClient.getConfig();
   * console.log(`IMAP enabled: ${config.imap_enabled}`);
   * console.log(`Sync interval: ${config.sync_interval_minutes} minutes`);
   * console.log(`Folders: ${config.sync_folders.join(', ')}`);
   * ```
   */
  async getConfig(): Promise<EmailConfig> {
    try {
      const response: AxiosResponse<EmailConfig> = await this.client.get(
        '/api/email-sync/config'
      );
      return response.data;
    } catch (error) {
      throw this.transformError(error);
    }
  }

  /**
   * Update email configuration
   *
   * Updates the email configuration for IMAP/SMTP synchronization.
   * Passwords can be updated but are never returned in the response.
   *
   * Only the fields specified in the request will be updated. All other fields
   * will remain unchanged. Partial updates are supported.
   *
   * @param request - Configuration update with only the fields to change
   * @returns Updated configuration (excluding passwords)
   * @throws ApiError if update fails
   *
   * @example
   * ```ts
   * // Enable IMAP and configure server settings
   * const updated = await emailSyncClient.updateConfig({
   *   imap_enabled: true,
   *   imap_server: 'imap.gmail.com',
   *   imap_port: 993,
   *   imap_use_ssl: true,
   *   imap_username: 'recruiting@example.com',
   *   imap_password: 'secure_password',
   *   sync_interval_minutes: 30
   * });
   *
   * // Update just the sync interval
   * const intervalUpdate = await emailSyncClient.updateConfig({
   *   sync_interval_minutes: 60
   * });
   * ```
   */
  async updateConfig(request: EmailConfigUpdate): Promise<EmailConfig> {
    try {
      const response: AxiosResponse<EmailConfig> = await this.client.put(
        '/api/email-sync/config',
        request
      );
      return response.data;
    } catch (error) {
      throw this.transformError(error);
    }
  }

  /**
   * Get the underlying Axios instance
   *
   * This is useful for making custom requests not covered by the convenience methods.
   *
   * @returns Axios instance
   */
  getAxiosInstance(): AxiosInstance {
    return this.client;
  }
}

/**
 * Default email sync client instance
 *
 * Use this singleton instance for all email sync calls.
 */
export const emailSyncClient = new EmailSyncClient();

/**
 * Export email sync client class for custom instances
 */
export default EmailSyncClient;
